### PR TITLE
Add variable query for distinct event property values

### PR DESCRIPTION
**Pull request by vereecw**

### CODE REVIEW

**Review by diericd — CHANGES_REQUESTED** [review]

File: src/QueryEditor/Events.tsx [hunk 1]
```diff
@@ -254,6 +264,14 @@ export const Events = (props: Props): JSX.Element => {
     props.onChangeEventQuery(updatedQuery)
   }
 
+  const getValueFilterOperators = (): KnownOperator[] => {
+    const operators: KnownOperator[] = ['=', '!=', '<', '<=', '>', '>=']
+    if (props.historianInfo && !semverCompare(props.historianInfo.Version, 'v7.3.0')) {
```
[the remark this answers is not in the document]
Comment (diericd) [reply src/QueryEditor/Events.tsx:269]: I'll put this in the other MR

File: pkg/api/util.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	}
+
+	for _, eventType := range eventTypes {
+		if eventType.UUID.String() == searchValue || eventType.Name == searchValue {
+			return map[uuid.UUID]schemas.EventType{eventType.UUID: eventType}
+		}
+	}
+
+	return filteredEventTypes
+}
```
[the remark this answers is not in the document]
Comment (vereecw) [reply pkg/api/util.go:136]: I think it wont make it much more readable

File: pkg/api/util.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+		for _, eventType := range eventTypes {
+			if eventType.UUID.String() == searchValue || eventType.Name == searchValue {
+				filteredEventTypes = append(filteredEventTypes, eventType)
+				break
+			}
+		}
+	}
+
+	return filteredEventTypes
+}
```
Comment (diericd) [inline pkg/api/util.go:128]: I think code deduplication trumps readability in this case.



```suggestion
func filterItems[T any](items []T, searchValue string, matchFuncs ...func(T) string) []T {
    filteredItems := make([]T, 0, len(items))
    if len(searchValue) == 0 {
        return filteredItems // Early exit for empty search
    }

    var re *regexp.Regexp
    if strings.HasPrefix(searchValue, "/") && strings.HasSuffix(searchValue, "/") {
        if len(searchValue) > 2 {
            pattern := searchValue[1 : len(searchValue)-1]
            var err error
            re, err = regexp.Compile(pattern)
            if err != nil {
                return filteredItems
            }
        } else {
            return filteredItems
        }
    }

    
    for _, item := range items {
        for _, matchFunc := range matchFuncs {
            if (re != nil && re.MatchString(matchFunc(item))) || (re == nil && matchFunc(item) == searchValue) {
                filteredItems = append(filteredItems, item)
                break
            }
        }
    }
    return filteredItems
}

func filterAssetUUIDs(assets []schemas.Asset, searchValue string) []schemas.Asset {
    return filterItems(assets, searchValue, 
        func(asset schemas.Asset) string { return asset.AssetPath },
        func(asset schemas.Asset) string { return asset.UUID.String() })
}

func filterEventTypeUUIDs(eventTypes []schemas.EventType, searchValue string) []schemas.EventType {
    return filterItems(eventTypes, searchValue, 
        func(eventType schemas.EventType) string { return eventType.Name },
        func(eventType schemas.EventType) string { return eventType.UUID.String() })
}
```

**Review by diericd — CHANGES_REQUESTED** [review]

File: src/CustomVariableEditor/VariableEditor.tsx [hunk 4]
```diff
@@ -8,21 +8,58 @@ import { MeasurementFilterRow } from './MeasurementFilter'
...
+    const options = [
+      { label: 'Measurement', value: VariableQueryType.MeasurementQuery },
+      { label: 'Asset', value: VariableQueryType.AssetQuery },
+      { label: 'Event type', value: VariableQueryType.EventTypeQuery },
+      { label: 'Database', value: VariableQueryType.DatabaseQuery },
+      { label: 'Event type property', value: VariableQueryType.EventTypePropertyQuery },
+      { label: 'Asset property', value: VariableQueryType.AssetPropertyQuery },
+    ] as Array<SelectableValue<string>>
+
+    if (historianInfo !== undefined && !semverCompare(historianInfo.Version, 'v7.3.0')) {
```
Comment (diericd) [inline src/CustomVariableEditor/VariableEditor.tsx:54]: whoooo major  jump here.

**Review by diericd — CHANGES_REQUESTED** [review]

**Comment by diericd** [comment]
Add value filters

image

**Review by diericd — CHANGES_REQUESTED** [review]
It is missing a time filter

File: src/CustomVariableEditor/PropertyValuesFilter.tsx [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+    const filter: EventPropertyFilter[] = []
+    updatedTags.forEach((tag) => {
+      const dataType = getDatatype(tag.key)
+      let eventPropertyFilter: EventPropertyFilter = {
+        Property: tag.key,
+        Datatype: dataType,
+        Condition: tag.condition || '',
+        Operator: tag.operator || '=',
+        Value: tag.value,
+        ScopedVars: {},
```
Comment (diericd) [inline src/CustomVariableEditor/PropertyValuesFilter.tsx:132]: what is this doing here?

File: pkg/datasource/resource_handler.go [hunk 6]
```diff
@@ -144,6 +146,23 @@ func (ds *HistorianDataSource) CallResource(ctx context.Context, req *backend.Ca
...
+		if err := ds.Decoder.Decode(&request, url.Query()); err != nil {
+			return err
+		}
+
+		o, err := ds.API.GetDistinctEventPropertyValues(ctx, eventTypePropertyUUID, request)
+		if err != nil {
+			return err
+		}
+
+		return resource.SendJSON(sender, o)
```
Comment (diericd) [inline pkg/datasource/resource_handler.go:165]: This is missing the time filter

**Review by diericd — CHANGES_REQUESTED** [review]

**Review by diericd — CHANGES_REQUESTED** [review]

**Review by diericd — APPROVED** [review]

**Pull request merged**